File: python3/letter_combinations_of_a_phone_number.py
# Space	: O(n)
# Time		: O(n!)


class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        ans = []
        phone_map = {
            '2': ['a', 'b', 'c'],
            '3': ['d', 'e', 'f'],
            '4': ['g', 'h', 'i'],
            '5': ['j', 'k', 'l'],
            '6': ['m', 'n', 'o'],
            '7': ['p', 'q', 'r', 's'],
            '8': ['t', 'u', 'v'],
            '9': ['w', 'x', 'y', 'z'],
        }
        n = len(digits)
        
        if not digits:
            return ans
        
        def combination(digs, path):
            if len(path) == n:
                ans.append(path)
            if len(digs) > 0:
                for letters in phone_map[digs[0]]:
                    for letter in letters:
                        combination(digs[1:], path + letter)
        
        combination(digits, "")
        
        return ans

File: golang/letter_combinations_of_a_phone_number.go
package main

/*
Space	: O(n)
Time	: O(n ^ 3)
*/

func letterCombinations(digits string) []string {
	ans := []string{}

	if digits == "" {
		return ans
	}

	n := len(digits)
	path := ""
	hashmap := map[string]string{
		"2": "abc",
		"3": "def",
		"4": "ghi",
		"5": "jkl",
		"6": "mno",
		"7": "pqrs",
		"8": "tuv",
		"9": "wxyz",
	}

	var getCombination func(i int)
	getCombination = func(i int) {
		if i == n {
			ans = append(ans, path)
			return
		}

		for _, v := range hashmap[string(digits[i])] {
			path += string(v)
			getCombination(i + 1)
			path = path[:len(path)-1]
		}
	}

	getCombination(0)
	return ans
}
